### ISSUE TITLE
Revisit the design of `AssocMap.Map`

### ISSUE THREAD

**Issue by ana-pantilie** [IntersectMBO/plutus#5921]
The [AssocMap.Map](https://github.com/IntersectMBO/plutus/blob/bcce39200ff7f3ed7321465c79baed4aa87cb1ad/plutus-tx/src/PlutusTx/AssocMap.hs#L72) type is confusing, since a `Map` type is usually expected to provide a very similar interface to https://hackage.haskell.org/package/containers-0.7/docs/Data-Map-Strict.html and to have an optimal way of storing/looking up/dealing with collisions of keys. This type does not provide any of that, and many of the functions/instances associated with it don't even enforce that the resulting map is well-defined. Also https://github.com/IntersectMBO/plutus/blob/bcce39200ff7f3ed7321465c79baed4aa87cb1ad/plutus-tx/src/PlutusTx/AssocMap.hs#L244 is non-standard.

See https://github.com/IntersectMBO/plutus/pull/5701 for more context.

**Comment by effectfully** [comment]
@ana-pantilie do we still need to do something here?

**Comment by ana-pantilie** [comment]
No, we've documented it and there should not be any confusion anymore.
